tests/dialects/test_mysql.py

```diff
@@ -506,6 +506,17 @@ def test_mysql_time(self):
         self.validate_identity("FROM_UNIXTIME(a, b)")
         self.validate_identity("FROM_UNIXTIME(a, b, c)")
         self.validate_identity("TIME_STR_TO_UNIX(x)", "UNIX_TIMESTAMP(x)")
+        self.validate_all(
+            "SELECT TO_DAYS('2007-10-07')",
+            write={
+                "mysql": "SELECT TO_DAYS('2007-10-07')",
+                "presto": "SELECT DATE_DIFF('DAY', CAST('0000-01-01' AS DATE), '2007-10-07')",
```

Comment:
Is `'2007-10-07'` a valid value for the third argument? Trino seems to expect a `DATE` here:

```
trino> SELECT DATE_DIFF('DAY', CAST('0000-01-01' AS DATE), '2007-10-07');
Query 20230927_183328_00002_4y9yd failed: line 1:8: Unexpected parameters (varchar(3), date, varchar(10)) for function date_diff. Expected: date_diff(varchar(x), date, date), date_diff(varchar(x), timestamp(p), timestamp(p)), date_diff(varchar(x), timestamp(p) with time zone, timestamp(p) with time zone), date_diff(varchar(x), time(p), time(p)), date_diff(varchar(x), time(p) with time zone, time(p) with time zone)
SELECT DATE_DIFF('DAY', CAST('0000-01-01' AS DATE), '2007-10-07')

trino> SELECT DATE_DIFF('DAY', CAST('0000-01-01' AS DATE), CAST('2007-10-07' AS DATE));
 _col0  
--------
 733321 
(1 row)

Query 20230927_183408_00003_4y9yd, FINISHED, 1 node
Splits: 1 total, 1 done (100.00%)
0.54 [0 rows, 0B] [0 rows/s, 0B/s]
```

Reply:
Perhaps MySQL needs to parse `TO_DAYS`'s argument into a `TsOrDsToDate`?

Reply:
Yes, that's right.
I think this test just isn't right.

I added the type casting here:
https://github.com/tobymao/sqlglot/pull/2334/files#diff-d19f6f1f7cc671a042d531f94b85a280e7b62a4b43b1acc1d2b57270d03fbdafR357

I'm not sure of a better place to do it.

Reply:
That cast doesn't apply to `expression.this` though. If you pass in a string, the transform will use that in `DateDiff`. What I meant is to add an entry in `MySQL.Parser.FUNCTIONS` so that we wrap the argument of `TO_DAYS` with a `TsOrDsToDate`. Then the generation should just work for Presto.

Reply:
Ah, sorry. Wrong link.
https://github.com/tobymao/sqlglot/pull/2334/files#diff-28f5a8e638b2f114cfd7411b4589283d786380e7f1cc040fd990c8290887b2dcR62

We apply the cast to expression.this.

If we cast during parsing, then we must do this for _all_ date functions. Since, for example, MySQL almost always allows strings for date arguments, and Presto almost always doesnt.

That could lead to a lot of redundant casting.

Reply:
>  I think this test just isn't right.

It's just that the mysql -> presto transpilation wasn't working properly due to the cast being omitted; with the suggested change above I think you'll be able to add that test back again but with the cast added

Reply:
> If we cast during parsing, then we must do this for all date functions. Since, for example, MySQL almost always allows strings for date arguments, and Presto almost always doesnt.

Isn't the point to do that everywhere so that the transpilation can work without the optimizer though? For example we parse several date expressions in redshift like that.

Reply:
I need to essentially the same thing for everything here: https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html

Many of these are super common date functions. What's different is the _types_ that are allowed.

If we push type casting to the parser, then all users will likely end up with lots of redundant casts. For example:
`SELECT YEAR(CAST('2021-01-01' AS DATE))` -> `SELECT YEAR(CAST(CAST('2021-01-01' AS DATE) AS DATE))`

And this will apply to _all_ dialects, not just mysql.

We can move the casting to the generator, in which case this redundancy will only be the case for certain dialects (e.g. presto).

Or, we can keep it in the optimizer, in which case we can be smart about when to add the cast, but transpiling won't automatically work in some cases.

The optimizer feels best to me, but maybe that's because I'm already using it.

Reply:
I get where you're coming from. Maybe another idea is to both handle this at parse time and try to reduce the redundant casts at generation time by applying the `remove_redundant_casts` rule after running the type annotations. Although, that does add some more overhead to the generator.

@tobymao what do you think?

Reply:
i wouldn't worry about redundant casts for now, but for this specific case, i think to_days is mysql specific enough that it may make sense to do this in the generator

Reply:
moving this to tsordstodate during parsing...